usim/py/events.py

```diff
@@ -0,0 +1,642 @@
+            "Test that at least 50% of events have triggered"
+            return count * 2 >= len(events)
+
+        majority = Condition(most, events)
+
+    The value of a :py:class:`~.Condition` is a :py:class:`~.ConditionValue`
+    of **all** the events used in the condition. Note that events of nested
+    conditions are flattened. If *any* event fails, the :py:class:`~.Condition`
+    fails with the same reason.
+
+    .. note::
+
+        The condition is only evaluated on instantiation and when child events
+        trigger. This means that ``evaluate`` should not depend on external,
+        mutable objects.
+
+    .. hint::
+
+        **Migrating to μSim**
+
+        Since conditions, such as a :py:class:`~.Flag`, may change their value,
+        μSim does not support arbitrary comparisons. The common and well-defined
+        :py:meth:`~.all_of` and :py:meth:`~.any_of` correspond to the ``&`` and ``|``
+        operators.
+
+        .. code:: python3
+
+            all_events = flag1 & flag2 & flag3
+            any_events = flag1 | flag2 | flag3
+    """
+    __slots__ = '_evaluate', '_events', '_processed'
+
+    def __init__(self, env, evaluate, events: Iterable[Event]):
+        super().__init__(env)
+        self._evaluate = evaluate
+        self._events = tuple(events)
+        if any(event.env != env for event in self._events):
+            raise ValueError('Events from multiple environments cannot be mixed')
+        env.schedule(self._check_events(), delay=0)
+
+    async def _check_events(self):
+        observed = [event for event in self._events if event.__usimpy_flag__]
+        unobserved = [event for event in self._events if not event.__usimpy_flag__]
+        for event in observed:
+            if not event.ok:
+                # TODO: taken from simpy - this might swallow exceptions
+                #       if event is awaited but self is not awaited
+                event.defused = True
+                self.fail(event.value)
+                return
+        while unobserved and not self._evaluate(self._events, len(observed)):
+            await AnyFlag(*(event.__usimpy_flag__ for event in unobserved))
+            for event in unobserved[:]:
+                if not event.__usimpy_flag__:
+                    continue
+                elif event.ok:
+                    unobserved.remove(event)
+                    observed.append(event)
+                else:
+                    event.defused = True
+                    self.fail(event.value)
+                    return
+        if self._evaluate(self._events, len(observed)):
+            self.succeed(ConditionValue(*self._flatten_values(self._events)))
+
+    @classmethod
+    def _flatten_values(cls, events) -> List[Event]:
+        """Flatten the values of all events"""
+        result = []
+        for event in events:
+            if isinstance(event, Condition):
+                result.extend(cls._flatten_values(event._events))
+            elif event.ok:
```

Comment:
Why is here the distinction between `Condition` and `event.ok`, what about `event.fail`? (I suppose a failure is also a kind of value)

Reply:
For a Condition, we *must* recurse to flatten it - this is because ``|`` and ``&`` produce a tree, not a chain. (See above comment why my attempts to fix this failed miserably.) If any event is ``event.fail``, then the entire Condition has failed already.

I guess an ``assert`` may be appropriate, both for documentation and robustness.

Reply:
So I tried to come up with *some* paper assertion here but I did not find one. The publicAPI of ``Event`` is surprisingly bare. What could work is this:

```python3
assert event._value is None or event._value[1] is None or event.defused
```
Not sure if that is helpful at this point.